Add support for different LMS levels to use different parameter sets

fluppe2 opened the pull request:
The interface to HSS now handles single type codes (as previously) and also lists of type codes. Using lists it is now possible to instantiate a heterogeneous HSS instance as allowed in RFC8554 (see §6).

fluppe2 committed:
Add support for different LMS levels to use different parameter sets

russhousley commented:
I think that the checking of lmots_type and lms_type has been lost.  I think we still want to raise an error if each value in the list is not known:

            raise ValueError(err_unknown_typecode, toHex(lmots_type))

Likewise, when converting a single value to a list, that should raise an error if the provided lmots_type and lms_type is unknown.

fluppe2 commented:
Hi Russ,

that should be present in the classes "HssPrivateKey" and 
"HssLmsPrivateKey" as follows (did you try with bogus typecodes?):

==========================================

In "HssPrivateKey" I replaced your previous lines:

if lmots_type not in lmots_params:
     raise ValueError(err_unknown_typecode, toHex(lmots_type))
if lms_type not in lms_params:
     raise ValueError(err_unknown_typecode, toHex(lms_type))

by checks on the whole lists

if not all(item in lmots_params for (item) in lmots_type):
     raise ValueError(err_unknown_typecode, str([toHex(item) for item in 
lmots_type]))
if not all(item in lms_params for (item) in lms_type):
     raise ValueError(err_unknown_typecode, str([toHex(item) for item in 
lms_type]))

==========================================

In "HssLmsPrivateKey" I replaced your previous lines:

if not lms_type in lms_params:
      raise ValueError(err_unknown_typecode)
if not lmots_type in lmots_params:
     raise ValueError(err_unknown_typecode)

by the check on the whole list

if not all(item in lms_params for (item) in lms_type):
     raise ValueError(err_unknown_typecode)
if not all(item in lmots_params for (item) in lmots_type):
     raise ValueError(err_unknown_typecode)

==========================================

In "HssLmsPrivateKey" your raise didn't print the typecodes, i.e. there 
was "raise ValueError(err_unknown_typecode)", so I took this unchanged 
and only replaced the condition to work on lists.

I chose to convert single values first to a list and then to check the 
list to have cleaner code. But certainly I can add another check on 
single values before converting.

If I've missed the place you spotted something missing, can you comment 
the specific code in the PR on github or maybe give me the lines of 
code/class/method you are referring to?

Best
Stavros

On 1/25/23 23:03, Russ Housley wrote:
>
> I think that the checking of lmots_type and lms_type has been lost. I 
> think we still want to raise an error if each value in the list is not 
> known:
>
> |raise ValueError(err_unknown_typecode, toHex(lmots_type)) |
>
> Likewise, when converting a single value to a list, that should raise 
> an error if the provided lmots_type and lms_type is unknown.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/russhousley/pyhsslms/pull/4#issuecomment-1404284614>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AC4UJ4ZJTQDTHIJAPYSLFFLWUGPJZANCNFSM6AAAAAAUC2SLWM>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

fluppe2 committed:
Create python-app.yml

russhousley commented:
I want this code to work in Python 2.7, even though it is obsolete.  I get a run-time error:

~~~
SyntaxError: Non-ASCII character '\xc2' in file pyhsslms/pyhsslms.py on line 998, but no encoding declared; see http://python.org/dev/peps/pep-0263/ for details
~~~

This is actually the use of non-ASCII characters in a comment.

russhousley commented:
The change that you made in TestHSS means that this change is not backward compatible.  That is, if a private key was generated with the current code, then it cannot be used to sign with the new code.

WeboFingerprints commented:
Hi Russ,

Is this compatible with the LMS crypto signature in mbeldTLS 3.6.1? I am using pyhsslms.HssLmsPrivateKey.sign() with level 1 to get a signature of 1456 bytes and the public keys. But I cannot get the signature to be verified with mbedtls_lms_verify() after imports the public keys and signature in Mbedtls.
